Hotfix for PR #7778 on Silicon Labs targets

stevew817 opened the pull request:
### Description

TB_SENSE_12 was left behind by the changes in #7778. As discussed with @SeppoTakalo over email, this commit implements the changes required to allow TB_SENSE_12 to provide a default network interface for Silicon Labs targets. This was done by mimicking the changes that were put in place by @SeppoTakalo for NCS36510/KW24D.

Considering #7778 made it in for 5.10.0-rc1, I strongly suggest you pull in this PR for the same release to keep continuity. CC @0xc0170 @kjbracey-arm @screamerbg 

### Pull request type

    [X] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Breaking change

stevew817 committed:
Hotfix for PR #7778

TB_SENSE_12 would have been left behind by the changes in #7778. This commit implements the changes in mbed to allow targets to provide a default network interface for Silicon Labs targets.

cmonr commented:
/morph build

cmonr commented:
@stevew817 We'll do what we can to get this in for RC2 (currently slipping to Monday), but it looks like this still needs some work with TB_Sense_1

stevew817 commented:
@SeppoTakalo TB_SENSE_1 is failing because some (netsocket) tests do not fit in its memory. What's the recommended course of action to turn off these tests?

0xc0170 commented:
> @SeppoTakalo TB_SENSE_1 is failing because some (netsocket) tests do not fit in its memory. What's the recommended course of action to turn off these tests?

@ARMmbed/mbed-os-test Please review

SeppoTakalo commented:
Recommended way is to have:
```
    "TB_SENSE_12": {
        "default_test_configuration": "NO_NETWORK",
```
in the test configuration, in case test cases do not fit.

But this is already provided, so I don't understand why it does not build...

SeppoTakalo commented:
```
$ mbed test --compile -n mbed-os-tests-netsocket\*
...
Compile [ 73.3%]: main.cpp
[Error] main.cpp@21,2: #error [NOT_SUPPORTED] No network configuration found for this target.
...
Build successes:
  * TB_SENSE_12::GCC_ARM::MBED-BUILD


Build skips:
  * TB_SENSE_12::GCC_ARM::MBED-OS-TESTS-NETSOCKET-DNS
  * TB_SENSE_12::GCC_ARM::MBED-OS-TESTS-NETSOCKET-TCP
  * TB_SENSE_12::GCC_ARM::MBED-OS-TESTS-NETSOCKET-UDP
```

This is what happens on my machine, so I don't understand why the skipping does not work on CI

0xc0170 commented:
Enable these `-DMBED_HEAP_STATS_ENABLED=1 -DMBED_STACK_STATS_ENABLED=1 -DMBED_TRAP_ERRORS_ENABLED=1 -DMBED_ALL_STATS_ENABLED` - might do the difference.

stevew817 commented:
Please keep in mind that it is `TB_SENSE_1` which is failing, not `TB_SENSE_12`. `TB_SENSE_1` is a target based on EFR32MG1P which has 256k flash/64k ram, so I wouldn't be surprised if the stack doesn't fit in there... However, it used to work with mesh-minimal-example, so...
@SeppoTakalo Any suggestions?

SeppoTakalo commented:
@stevew817 Please provide the test config for `TB_SENSE_1` as well, if both have 802.15.4 RF enabled.

Mesh application might fit, if configured to be minimal enough, but same does not apply to our test applications which might use full set of Socket operations, DNS libraries, etc.

So in this first phase, provide the support for interface with test configurations disabled.
Then later on we will work out how to get mesh boards into test.

Also, another reason to use the `NO_NETWORK` config in test, is that the CI that runs these PR tests, is not located in place where mesh network is available. We have separate CI for that.

So, add:
```
     },
    "TB_SENSE_1": {
        "default_test_configuration": "NO_NETWORK",
        "test_configurations": ["6LOWPAN_HOST", "6LOWPAN_ROUTER", "THREAD_END_DEVICE", "THREAD_ROUTER"]
    }
```
to the `target_configs.json` as well.

stevew817 committed:
Supply default test config for TB_SENSE_1 as well

stevew817 commented:
@SeppoTakalo @0xc0170 @cmonr fixed.

cmonr commented:
/morph build

stevew817 commented:
Unrelated failure...

cmonr commented:
Correct, both are not.
/morph test

stevew817 commented:
uvision is having a bad day?

cmonr commented:
Actually, we just had to restart our license server. Kinda embarassing really.

/morph export-build

cmonr commented:
@stevew817 I'm still looking into why the mbed2-build is not working.

cmonr commented:
Well that's a new type of failure:
```
16:32:37 FATAL: Channel "unknown": Remote call on k64f_aws_build_beast failed. The channel is closing down or has closed down
```

/morph test

@ARMmbed/mbed-os-test Fyi ^^^

cmonr commented:
Rerunning just to check, but I think this is still an issue.

 /morph mbed2-build

cmonr commented:
/morph mbed2-build

kjbracey commented:
/morph mbed2-build

kjbracey commented:
The failure appears to be something real hitting just this target, but not getting much in way of diagnostics:

```
Building library CMSIS (TB_SENSE_12, IAR)
Macros: -DDEVICE_CRC=1 -DDEVICE_SPI=1 -D__MBED__=1 -DDEVICE_I2CSLAVE=1 -DMBED_BUILD_TIMESTAMP=1536568849.82 -DDEVICE_PORTINOUT=1 -DTARGET_RTOS_M4_M7 -DDEVICE_RTC=1 -DDEVICE_SERIAL_ASYNCH=1 -D__CMSIS_RTOS -DTRANSACTION_QUEUE_SIZE_SPI=4 -DMBEDTLS_CONFIG_HW_SUPPORT -DDEVICE_SPISLAVE=1 -DDEVICE_USTICKER=1 -DDEVICE_STDIO_MESSAGES=1 -DTARGET_EFR32MG12P332F1024GL125 -DTARGET_CORTEX_M -DDEVICE_I2C_ASYNCH=1 -DTARGET_LIKE_CORTEX_M4 -DTARGET_M4 -DARM_MATH_CM4 -DEFR32MG12P332F1024GL125 -DDEVICE_SPI_ASYNCH=1 -DDEVICE_LPTICKER=1 -DTOOLCHAIN_IAR -DMBED_TICKLESS -DDEVICE_INTERRUPTIN=1 -DTARGET_CORTEX -DDEVICE_I2C=1 -DDEVICE_PORTOUT=1 -D__CORTEX_M4 -D__MBED_CMSIS_RTOS_CM -DDEVICE_PWMOUT=1 -D__FPU_PRESENT=1 -DDEVICE_PORTIN=1 -DTARGET_RELEASE -DDEVICE_802_15_4_PHY=1 -DTARGET_EFM32 -DTARGET_EFR32MG12 -DEM_MSC_RUN_FROM_FLASH -DDEVICE_TRNG=1 -DTARGET_LIKE_MBED -DTARGET_TB_SENSE_12 -DDEVICE_SLEEP=1 -DTARGET_EFR32_12 -DTARGET_Silicon_Labs -DTARGET_SL_CRYPTO -DDEVICE_ANALOGIN=1 -DDEVICE_SERIAL=1 -DTARGET_SL_RAIL -DDEVICE_FLASH=1 -DTARGET_1024K
Compile [ 50.0%]: cmain.S
Compile: /usr/local/iccarm-7.80.1/bin/iasmarm --cpu Cortex-M4F -D__CORTEX_M4 -DARM_MATH_CM4 -D__FPU_PRESENT=1 -DEM_MSC_RUN_FROM_FLASH -D__MBED_CMSIS_RTOS_CM -DEFR32MG12P332F1024GL125 -D__CMSIS_RTOS -DTRANSACTION_QUEUE_SIZE_SPI=4 -DMBED_TICKLESS -DMBEDTLS_CONFIG_HW_SUPPORT -f /builds/ws/mbed-os2-build-matrix/1559/TB_SENSE_12_IAR/sources/mbed-os/BUILD/mbed/.temp/TARGET_TB_SENSE_12/TOOLCHAIN_IAR/asm_macros_9594ea6abafd3b9666368025b40c38b5.xcl -o BUILD/mbed/.temp/TARGET_TB_SENSE_12/TOOLCHAIN_IAR/cmsis/TARGET_CORTEX_M/TOOLCHAIN_IAR/cmain.o /builds/ws/mbed-os2-build-matrix/1559/TB_SENSE_12_IAR/sources/mbed-os/cmsis/TARGET_CORTEX_M/TOOLCHAIN_IAR/cmain.S
[DEBUG] Return: -11
```

stevew817 commented:
@kjbracey-arm Have you seen this before? I tried searching for this '-11' error code, but came up blank. Additionally, it's a CMSIS file, so shouldn't it 'just work'?

kjbracey commented:
I'm at something of a loss I'm afraid - not seen this before. Only thing I can guess is overrunning a command-line length limit or something, but I don't know if that makes sense - no other reason to suppose this particular target is at the limit. Will try to ask around.

adbridge commented:
@theotherjimmy We have trouble compiling TB_SENSE_12 in the online IDE for mbed2 and have done for some time due to the the IDE returning unknown target. Could this be related ?

cmonr commented:
@studavekar @ARMmbed/mbed-os-test I'm having trouble reproducing the issue locally, heck even reproducing the same CI output. Simply running the command at the top of the log file does not result in the same output.

Thoughts? Some help?

cmonr commented:
/morph mbed2-build

stevew817 commented:
Would be nice to get something to reproduce and figure out why this is failing. @cmonr any luck?

cmonr commented:
@stevew817 Sorry, but I haven't had much luck reproducing the problem.

I have a hunch, but I'm not completely sure if it will work.. Could you create a new PR with the same commits?

I have a suspicion that if the build path changes, the problem will go away. It doesn't make any sense at all, but with an earlier PR, we saw similar behavior, and that seemed to work.

cmonr commented:
Trying a thing. As of right now this is good to go aside for the strange ci-morph-mbed2-build build issue.

NirSonnenschein commented:
/morph mbed2-build

kjbracey commented:
-11 failure presumably related to #8017 and problem found in #7662 - some sort of malformed assembler macro, due to error processing JSON.

Need to look to see if there's something unpleasant in the generated `.xcl` macro file.

Also, maybe something to do with `TB_SENSE_12` being misinterpreted as `TB_SENSE_1` + `2`?

cmonr commented:
/morph mbed2-build

cmonr commented:
/morph mbed2-build

cmonr commented:
Depends on this fix: https://github.com/ARMmbed/mbed-os/pull/8161

0xc0170 commented:
/morph mbed2-build

0xc0170 commented:
I downloaded artifacts from the latest build.

Here's asm.xcf file
```
"-DMBED_CONF_TARGET_NETWORK_DEFAULT_INTERFACE_TYPE=MESH" "-DHFXO_FREQUENCY=38400000" "-DLFXO_FREQUENCY=32768" "-DMBED_CONF_SL_RAIL_PTI_MODE=RAIL_PTI_MODE_UART" "-DMBED_CONF_SL_RAIL_PTI_BAUDRATE=1600000" "-DMBED_CONF_PLATFORM_ERROR_HIST_ENABLED=0" "-DMBED_CONF_SL_RAIL_PTI_DOUT_LOCATION=6" "-DMBED_CONF_PLATFORM_ERROR_DECODE_HTTP_URL_STR=\nFor more info, visit: https:/\/armmbed.github.io/mbedos-error/?error=0x%08X" "-DMBED_CONF_SL_RAIL_PTI_DCLK_PORT=gpioPortB" "-DMBED_CONF_PLATFORM_FORCE_NON_COPYABLE_ERROR=0" "-DHFRCO_FREQUENCY_ENUM=cmuHFRCOFreq_32M0Hz" "-DMBED_CONF_SL_RAIL_PTI_DOUT_PORT=gpioPortB" "-DMBED_CONF_DRIVERS_UART_SERIAL_RXBUF_SIZE=256" "-DMBED_CONF_SL_RAIL_PTI_DFRAME_PIN=13" "-DMBED_CONF_SL_RAIL_PTI=1" "-DMBED_MEM_TRACING_ENABLED=0" "-DMBED_CONF_SL_RAIL_PTI_DFRAME_PORT=gpioPortB" "-DMBED_CONF_PLATFORM_POLL_USE_LOWPOWER_TIMER=0" "-DMBED_CONF_PLATFORM_STDIO_FLUSH_AT_EXIT=1" "-DMBED_CONF_PLATFORM_STDIO_BAUD_RATE=115200" "-DMBED_CONF_SL_RAIL_HAS_SUBGIG=0" "-DMBED_CONF_SL_RAIL_HAS_2P4=1" "-DMBED_CONF_PLATFORM_STDIO_BUFFERED_SERIAL=0" "-DMBED_CONF_SL_RAIL_PTI_DOUT_PIN=12" "-DMBED_CONF_SL_RAIL_BAND=2400" "-DMBED_CONF_SL_RAIL_PTI_DFRAME_LOCATION=6" "-DMBED_CONF_SL_RAIL_PTI_DCLK_PIN=11" "-DMBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE=9600" "-DMBED_CONF_SL_RAIL_PTI_DCLK_LOCATION=6" "-DMBED_CONF_PLATFORM_ERROR_ALL_THREADS_INFO=0" "-DCORE_CLOCK_SOURCE=HFXO" "-DMBED_CONF_PLATFORM_MAX_ERROR_FILENAME_LEN=16" "-DMBED_CONF_PLATFORM_ERROR_HIST_SIZE=4" "-DMBED_CONF_DRIVERS_UART_SERIAL_TXBUF_SIZE=256" "-DLOW_ENERGY_CLOCK_SOURCE=LFXO" "-DMBED_CONF_PLATFORM_STDIO_CONVERT_TTY_NEWLINES=0" "-DHFRCO_FREQUENCY=32000000" "-DMBED_CONF_PLATFORM_ERROR_FILENAME_CAPTURE_ENABLED=0" "-DMBED_CONF_PLATFORM_STDIO_CONVERT_NEWLINES=0"
```

the url seems to be correct (the IAR fix is in place).

profile-asm:

```

{
    "flags": [], 
    "macros": [], 
    "symbols": [
        "ARM_MATH_CM4", 
        "EFR32MG12P332F1024GL125", 
        "EM_MSC_RUN_FROM_FLASH", 
        "MBEDTLS_CONFIG_HW_SUPPORT", 
        "MBED_TICKLESS", 
        "TRANSACTION_QUEUE_SIZE_SPI=4", 
        "__CMSIS_RTOS", 
        "__CORTEX_M4", 
        "__FPU_PRESENT=1", 
        "__MBED_CMSIS_RTOS_CM"
    ]
}
```

kjbracey commented:
Could there be a more general string-handling problem here - I note that the IAR conversion has always stripped the `"` characters from the defines altogether. Doing that leaves something that seems a bit iffy, including in that URL case, and I doubt such a mangled define would ever be useful to the assembler. Should it not be replacing `"` with `\"`? Or if it can't, would be better to omit the untranslatable defines fully.

0xc0170 commented:
I locally build this PR. Compared the files to the artifacts from the failing build - they match 99 percent (only diff is the timestamp macro). Locally this builds, using `IAR ANSI C/C++ Compiler V7.80.4.12462/W32 for ARM`.

kjbracey commented:
Getting a bit desperate here. Upgrade the CI to use newer toolchain? Why is it on 7.80.1?

I can't see any reason this failure should be specific to the new target being added here - I see nothing weird in the xcl due to this target.

All we're left with is the anthropomorphic principle - it can only fail this target because we're in the timeline where all other targets are already passing.

So, stretching a bit, I'm assuming that unquoted URL mess in the defines is still problematic for the assembler, and somehow this new target has tripped something by alignment.

Now, we are not passing valid strings in, because of the stripping of `"`. Therefore attempts to mangle `//`, which must be in a string, is pointless.   We're only side-stepping the bug, not providing a useful definition to the assembler.

Let's try change the transformation to just totally remove all defines containing `"` from the list, given that we know no-one can be using them in assembler.  That would get rid of the nastiest thing in the xcl file.

kjbracey commented:
I don't speak Python, but I think that's something like:

            macros_cmd = ['"-D%s"' % d for d in config_macros if not '"' in d]

Shall we try adding that to this PR?

0xc0170 commented:
> Let's try change the transformation to just totally remove all defines containing " from the list, given that we know no-one can be using them in assembler. That would get rid of the nastiest thing in the xcl file.

@ARMmbed/mbed-os-tools

kjbracey commented:
Alternatively, of course, figure out how to escape the `"` properly. I have no idea of the history behind the stripping there. With the quotes in, the `//` would probably not be a problem. Without them the `//` is inherently problematic, because that makes it a macro containing a line comment.

And that _is_ something the assembler conceptually permits, unlike the C compiler - there are grave warnings against carelessly doing `#define foo x // blah` in the manual, as that will define foo as `x // blah`.

kjbracey commented:
Did @cmonr actually find that #8161 - removing the "//" at source - would make this pass on the CI machine?

kjbracey commented:
Right, the quote stripping is actually fairly recent: https://github.com/ARMmbed/mbed-os/pull/7061#issuecomment-398332982

(Prior to that PR the macros weren't being passed at all)

I'll try experimentally just dropping any macros with quotes, but maybe replacing with `\"` would have worked, @TTornblom - did you try this?

kjbracey committed:
Remove IAR assembler macros containing quotes

IAR assembler 7.80 has some problems handling difficult macros, leading
to immediate exit with return value -11.

In particular, a URL string has been causing problems, presumably due to
the "//" resembling a comment.

A previous escaping workaround in 0d97803 seemed to work, but the crash
has still been seen with a particular target.

Previous creation of the extended command line file for the IAR
assembler was stripping quotes from macros. This rendered the resulting
definitions for string-containing macros incorrect, which means that we
can assume no assembler code is currently relying on them.

Therefore, as a precautionary measure to avoid the crash, simply remove
all macros containing strings when creating them for IAR. This
apparently clears the crashes seen during testing of
https://github.com/ARMmbed/mbed-os/pull/8023

kjbracey commented:
/morph mbed2-build

kjbracey commented:
Okay, dropping macros with quotes cleared the test.

stevew817 commented:
This can't be the only target that is suffering from such a generic failure mode???

kjbracey commented:
> This can't be the only target that is suffering from such a generic failure mode???

At the minute, it does seem to be. I think you've been spectacularly unlucky here.

The fundamental problem in question is fairly recent - apparently an IAR assembler bug triggered by the complicated (and wrongly-quoted) new error message config option that's recently been added to JSON.

There must be some lingering instability not addressed by the previous workaround, and somehow your xcl file manages to achieve some sort of failure (on 7.80.1, but not 7.80.4?). Presumably the same thing would pop up on other targets later if they changed something about their config - it seems deterministic for a particular config.

I've done a bit more local testing - it seems that `""` is the correct substitution to get a quote into the assembler to define a string, so `"-Dstr=""hello"""`. However, that doesn't stop `"-Dstr=""http://"""` from exploding 7.80.2. Substituting that to make `"-Dstr=""http:/\/"""` does feed in the correct string with no crash, at the cost of an "unrecognised escape sequence" warning if the string is used.

However, we still don't know what exactly it is that is upsetting just this particular target. I will try that substitution, to see if maybe adding the quotes back avoids some other problem with the `\n` or something. (Not that \n is a valid escape for the assembler either...)

But given that we already know the -11 error is target-specific, maybe /any/ change we made to the xcl file would make it go away like for the other targets. It seems like the safest bet is to err on the side of caution and eliminate all string macros, hoping and assuming that this -11 is also something about string macros.

kjbracey commented:
/morph mbed2-build

kjbracey commented:
And, that fails. Okay, should I make a separate PR for the string macro strip out, or fold it into this one?

It's logically separate, but CI load is high, and there's already loads of discussion here.

0xc0170 commented:
> It's logically separate, but CI load is high, and there's already loads of discussion here.

In this case here to make it green

kjbracey commented:
@theotherjimmy and @TTornblom could you review the last commit messing with IAR assembler, please?

0xc0170 commented:
/morph mbed2-build

TTornblom commented:
I have been playing some with the macro rewrite, and here is an alternative to remove the double quotes that seems to work. It will leave the double quotes in, but "//" still needs to be rewritten as "/\\/". 

```diff
diff --git a/tools/toolchains/iar.py b/tools/toolchains/iar.py
index b54487869..ed93dee1c 100644
--- a/tools/toolchains/iar.py
+++ b/tools/toolchains/iar.py
@@ -167,7 +167,7 @@ class IAR(mbedToolchain):
         opts = ['-D%s' % d for d in defines]
         if for_asm:
             config_macros = self.config.get_config_data_macros()
-            macros_cmd = ['"-D%s"' % d.replace('"', '').replace('//','/\/') for d in config_macros]
+            macros_cmd = ['"-D%s"' % d.replace('"', '""').replace('//','/\/') for d in config_macros]
             if self.RESPONSE_FILES:
                 via_file = self.make_option_file(
                     macros_cmd, "asm_macros_{}.xcl")
```

cmonr commented:
The plan:

I'm going to take @TTornblom's patch and commit it to this PR. If the mbed2-build job passes with it, then we'll progress the PR. If not, then I'll revert the PR, progress the PR asis, and open a new issue to look into why the patch didin't work.

cmonr commented:
/morph mbed2-build

cmonr commented:
Welp, since that didn't work, backing out the commit and progressing the PR with only @kjbracey-arm's change.

cmonr commented:
/morph mbed2-build

cmonr commented:
/morph build

0xc0170 commented:
/morph test

cmonr commented:
Phew. That took waaay too long.
